### PR TITLE
Added Order and OrderItems entity.

### PR DESCRIPTION
**Pull request by sovisou**

### CODE REVIEW

File: src/main/java/com/onlinebookstore/controller/OrderController.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    public List<OrderDto> getAllOrders(Authentication authentication) {
+        Long userId = extractUserId(authentication);
+        return orderService.findAllOrders(userId);
+    }
+
+    @PatchMapping("/{id}")
+    @Operation(summary = "Update order status")
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    public OrderDto updateOrderStatus(@PathVariable Long orderId,
+                                      @RequestBody OrderUpdateDto updateDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/controller/OrderController.java]:
```suggestion
                                      @RequestBody @Valid OrderUpdateDto updateDto) {
```

File: src/main/java/com/onlinebookstore/controller/OrderController.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/orders")
+public class OrderController {
+    private final OrderService orderService;
+
+    @PostMapping
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/controller/OrderController.java:30]: add PreAuthorize where is needed

File: src/main/java/com/onlinebookstore/dto/order/OrderDto.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+@Data
+public class OrderDto {
+    private Long id;
+    private Long userId;
+    private Status status;
+    private BigDecimal total;
+    private LocalDateTime orderDate;
+    private String shippingAddress;
+    private Set<OrderItemDto> orderItems;
+
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/dto/order/OrderDto.java]:
```suggestion
```

File: src/main/java/com/onlinebookstore/dto/order/CreateOrderRequestDto.java [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package com.onlinebookstore.dto.order;
+
+import jakarta.validation.constraints.NotBlank;
+import lombok.Data;
+
+@Data
+public class CreateOrderRequestDto {
+    @NotBlank
+    private String shoppingAddress;
+
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/dto/order/CreateOrderRequestDto.java]:
```suggestion
```

File: src/main/java/com/onlinebookstore/dto/order/OrderUpdateDto.java [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package com.onlinebookstore.dto.order;
+
+import com.onlinebookstore.enums.Status;
+import jakarta.validation.constraints.NotBlank;
+import lombok.Data;
+
+@Data
+public class OrderUpdateDto {
+    @NotBlank
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/dto/order/OrderUpdateDto.java]:
```suggestion
    @NotNull
```

File: src/main/java/com/onlinebookstore/service/impl/OrderServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+                    orderItem.setQuantity(cartItem.getQuantity());
+                    orderItem.setOrder(order);
+                    return orderItem;
+                })
+                .collect(Collectors.toSet());
+        order.setOrderItems(orderItems);
+        order.setTotal(orderItems.stream()
+                .map(item -> item.getPrice().multiply(BigDecimal.valueOf(item.getQuantity())))
+                .reduce(BigDecimal.ZERO, BigDecimal::add));
+        Order savedOrder = orderRepository.save(order);
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/service/impl/OrderServiceImpl.java]:
```suggestion
        orderRepository.save(order);
```

File: src/main/java/com/onlinebookstore/service/impl/OrderServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        return orderMapper.toDto(savedOrder);
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/service/impl/OrderServiceImpl.java]:
```suggestion
        return orderMapper.toDto(order);
```

File: src/main/java/com/onlinebookstore/service/impl/OrderServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    private final OrderItemMapper orderItemMapper;
+
+    @Override
+    @Transactional
+    public OrderDto createOrder(Long userId, CreateOrderRequestDto requestDto) {
+        User user = userRepository.findById(userId).orElseThrow(
+                () -> new EntityNotFoundException("Can't find user with id: " + userId));
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId).orElseThrow(
+                () -> new EntityNotFoundException("Can't find shopping cart for user with id: "
+                        + userId));
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/service/impl/OrderServiceImpl.java:48]: check if shop cart  `getCartItems().isEmpty()`

File: src/main/java/com/onlinebookstore/service/impl/OrderServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+        Order order = new Order();
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/service/impl/OrderServiceImpl.java]: divide logic to separate methods

File: src/main/java/com/onlinebookstore/service/impl/OrderServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+        order.setTotal(orderItems.stream()
+                .map(item -> item.getPrice().multiply(BigDecimal.valueOf(item.getQuantity())))
+                .reduce(BigDecimal.ZERO, BigDecimal::add));
+        Order savedOrder = orderRepository.save(order);
+        return orderMapper.toDto(savedOrder);
+    }
+
+    @Override
+    @Transactional
+    public List<OrderDto> findAllOrders(Long userId) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/service/impl/OrderServiceImpl.java]:
```suggestion
    public List<OrderDto> findAllOrders(Long userId, Pageable pageable) {
```

File: src/main/java/com/onlinebookstore/service/impl/UserServiceImpl.java [hunk 11]
```diff
@@ -38,7 +38,8 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto)
         user.setPassword(passwordEncoder.encode(requestDto.getPassword()));
         user.setRoles(Set.of(roleRepository.findByRole(Role.RoleName.ROLE_USER)
                 .orElseThrow(() -> new EntityNotFoundException("Role USER not found"))));
+        user = userRepository.save(user);
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/service/impl/UserServiceImpl.java]:
```suggestion
        userRepository.save(user);
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/com/onlinebookstore/dto/order/CreateOrderRequestDto.java [hunk 12]
```diff
@@ -0,0 +1,10 @@
+package com.onlinebookstore.dto.order;
+
+import jakarta.validation.constraints.NotBlank;
+import lombok.Data;
+
+@Data
+public class CreateOrderRequestDto {
+    @NotBlank
+    private String shoppingAddress;
```
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/dto/order/CreateOrderRequestDto.java:9]:
```suggestion
    private String shippingAddress;
```

File: src/main/java/com/onlinebookstore/service/impl/OrderServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+    @Override
+    @Transactional
+    public OrderDto createOrder(Long userId, CreateOrderRequestDto requestDto) {
+        User user = userRepository.findById(userId).orElseThrow(
+                () -> new EntityNotFoundException("Can't find user with id: " + userId));
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId).orElseThrow(
+                () -> new EntityNotFoundException("Can't find shopping cart for user with id: "
+                        + userId));
+        if (shoppingCart.getCartItems().isEmpty()) {
+            throw new EntityNotFoundException("Shopping cart is empty for user with id: " + userId);
```
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/service/impl/OrderServiceImpl.java:50]: better create another exception. For example OrderProcessingException (it should return status CONFLICT)

File: src/main/java/com/onlinebookstore/service/impl/OrderServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+        }
+        Order order = createOrderFromCart(user, shoppingCart, requestDto);
+        orderRepository.save(order);
+        return orderMapper.toDto(order);
+    }
+
+    @Override
+    @Transactional
+    public List<OrderDto> findAllOrders(Long userId, Pageable pageable) {
```
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/service/impl/OrderServiceImpl.java:59]: better return Page instead List

**Review by liliia-ponomarenko — APPROVED** [review]
Good job! A few minor comments. But please fix it before merging;)

**Pull request merged**